src/Playwright/Transport/WebSocketTransport.cs

```diff
@@ -0,0 +1,194 @@
+{
+    internal class WebSocketTransport : IConnectionTransport, IDisposable
+    {
+        private const int DefaultBufferSize = 1024;  // Byte buffer size
+        private readonly ClientWebSocket _webSocket;
+        private readonly string _wsEndpoint;
+        private readonly CancellationTokenSource _readerCancellationSource = new();
+        private readonly BrowserTypeConnectOptions _options;
+
+        internal WebSocketTransport(
+            string wsEndpoint = default,
+            BrowserTypeConnectOptions options = default)
+        {
+            _webSocket = new ClientWebSocket();
+            _wsEndpoint = wsEndpoint;
+            _options = options;
+            SetHeaders();
+            _ = ConnectAsync();
+        }
+
+        /// <inheritdoc cref="IDisposable.Dispose"/>
+        ~WebSocketTransport() => Dispose(false);
+
+        public event EventHandler<MessageReceivedEventArgs> MessageReceived;
+
+        public event EventHandler<LogReceivedEventArgs> LogReceived;
+
+        public event EventHandler<TransportClosedEventArgs> TransportClosed;
+
+        public bool IsClosed { get; private set; }
+
+        public async Task SendAsync(string message)
+        {
+            var messageBuffer = Encoding.UTF8.GetBytes(message);
+            var messagesCount = (int)Math.Ceiling((double)messageBuffer.Length / DefaultBufferSize);
+
+            try
+            {
+                if (!_readerCancellationSource.IsCancellationRequested)
+                {
+                    for (var i = 0; i < messagesCount; i++)
+                    {
+                        var offset = DefaultBufferSize * i;
+                        var count = DefaultBufferSize;
+                        var lastMessage = (i + 1) == messagesCount;
+
+                        if ((count * (i + 1)) > messageBuffer.Length)
+                        {
+                            count = messageBuffer.Length - offset;
+                        }
+
+                        await _webSocket.SendAsync(new ArraySegment<byte>(messageBuffer, offset, count), WebSocketMessageType.Text, lastMessage, _readerCancellationSource.Token).ConfigureAwait(false);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Close(ex);
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Close(string closeReason)
+        {
+            if (!IsClosed)
+            {
+                IsClosed = true;
+                _ = _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, closeReason, _readerCancellationSource.Token).ConfigureAwait(false);
+                TransportClosed?.Invoke(this, new TransportClosedEventArgs { CloseReason = closeReason });
+                _readerCancellationSource.Cancel();
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
```

Comment:
Why is this here?

Reply:
Freeing the WebSocket transport object once the job is done. Hence, required.

Reply:
It's a [common implementation](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#the-dispose-method) of `Dispose` in .NET. Basically, because we're calling `Dispose(true)` one line above, we call the `Suppress` to ensure we don't unnecessarily call the finalizer twice. The *actual* clean-up already happens in the first call, then we just clean the rest.

Reply:
So, suppress is required right? to clean the rest

Reply:
To clarify, `Suppress` doesn't clean, it prevents cleaning from happening *again* because the relevant (managed) finalization will happen in the `Dispose(true)` call. That's my interpretation at least.

Reply:
But we are not doing anything in this dispose, why do we even need it?

Reply:
He is: https://github.com/microsoft/playwright-dotnet/pull/1602/files/d10acf1b24cc4dc4fbdd8f94072b1597d99e64b4#diff-b29da783aeafa3800dd5b5b9d0c05f44cd36d1dcfc662527c209f55593aa9693R92

He's cancelling & disposing the cancellation source.

Reply:
Ah, this now huge inline comment confused me. Thanks!